Kube proxy rule reorg XLB->EXT

thockin opened the pull request:
This builds on #106497 and could sure use #107471 :)

/kind cleanup

```release-note
Users who look at iptables dumps will see some changes in the naming and structure of rules.
```

thockin committed:
kube-proxy: internal renames

thockin committed:
kube-proxy: Rename XLB -> EXT

This changes the "XLB" chain into the "EXT" chain - the "external
destinations" chain.

thockin committed:
kube-proxy: all external jumps to XLB chain

This makes the "destination" policy model clearer.  All external
destination captures now jump to the "XLB chain, which is the main place
that masquerade is done (removing it from most other places).

This is simpler to trace - XLB *always* exists (as long as you have an
external exposure) and never gets bypassed.

thockin committed:
kube-proxy: Rework LB VIP capture logic

* Comments
* If there are multiple VIPs, don't declare the fwChain multiple times.
* Don't emit the last -j DROP if there's no source ranges

thockin committed:
kube-proxy: Remove now unneeded rule

Now that NodePorts jump to EXT, we don't need a specific rule for
loopback source detection.

thockin committed:
kube-proxy: Renames for readability

thockin committed:
kube-proxy: remove old TODO

thockin committed:
kube-proxy: comment endpoint in SEP jumps

thockin committed:
kube-proxy: clean up tests

No functional changes, much whitespace.

Make assertIPTablesRulesEqual() *not* sort the `expected` value - make
the test cases all be pre-sorted.  This will make followup commits
cleaner.

Make the test output cleaner when this fails.

Use dedent everywhere for easier reading.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1042,20 +1046,30 @@ func (proxier *Proxier) syncProxyRules() {
+		// service-destined should jump.  ClusterIP traffic is considered
+		// "internal" while NodePort, LoadBalancer, and ExternalIPs traffic is
+		// considered "external".
+		internalTrafficChain := internalPolicyChain
+		externalTrafficChain := svcInfo.externalChainName // eventually jumps to externalPolicyChain
```
Having both `*policyChain` and `*trafficChain`, where `internalTrafficChain` is always the same as `internalPolicyChain`, but `externalTrafficChain` is not always the same as `externalPolicyChain` seems confusing.

I _think_ it would be clearer to have only 3 variables, corresponding to the things you call `internalPolicyChain`, `externalTrafficChain`, and `externalPolicyChain` here. But not with those exact names... It think the variable pointing to the EXT chain needs a name that makes it clearer that it's doing setup / preprocessing / special-casing. (And then I don't care much whether the other two have "Policy" or "Traffic" in their name.)

thockin replied:
I initially changed `*TrafficChain` to `*DestinationChain` so it was clearer that all of the `-j` code jumped to either `internalDestination` or `externalDestination` - but it was a lot more diff for less obvious value.

The extra level of indirection was intended to make clear that `policy{Cluster,Local}Chain` means "the chain to use for this policy" and `{internal,external}PolicyChain` means "which of cluster/local to use for internal/external", and `{internal,external}TrafficChain` was the indirection to hide the details of where to start.

Concretely, I though about adding an INT to match EXT, but INT is *always* empty.  

I also considered `policy{Cluster,Local}Chain` + `{itp,etp}Chain` + `{internal,external}DestChain`.  Or we could put this into a function like `internalDestChain, externalDestChain := chooseDestChains(svcInfo)`

I will look at this again.  If you're directionally OK with this PR, that's my main concern.

thockin replied:
I added a commit to add an INT but I don't think I like it perf-wise (but the symmetry is nice).

danwinship replied:
I don't think there's any reason to add an INT chain. My complaint was just that the naming was confusing, not that the organization of the chains was bad.

> I initially changed `*TrafficChain`

yeah, I think "Traffic" is the problem. Like, `externalPolicyChain` implements external policy (and `externalTrafficChain` does not). That makes sense. But both chains receive external traffic, so what exactly is it that `externalTrafficChain` does that `externalPolicyChain` doesn't?

That's why I was suggesting it needs a name like `externalTrafficSetupChain` or something... But then since there's no setup for internal traffic you don't need a corresponding extra variable for internal traffic.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1295,16 +1301,7 @@ func (proxier *Proxier) syncProxyRules() {
-						args,
-						"-j", string(KubeMarkMasqChain))
-				} else {
-					// TODO: Make all nodePorts jump to the firewall chain.
-					// Currently we only create it for loadbalancers (#33586).
```
if you're killing this comment it might be good to explain why in the commit message for future git archeology reference?

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1530,25 +1534,33 @@ func TestOverallIPTablesRulesWithMultipleServices(t *testing.T) {
+		-A KUBE-EXT-PAZTZYUUMV5KCDZL -j KUBE-SVC-PAZTZYUUMV5KCDZL
+		-A KUBE-EXT-X27LE4BHSL4DOUIK -m comment --comment "masquerade traffic for ns3/svc3:p80 external destinations" -j KUBE-MARK-MASQ
+		-A KUBE-EXT-X27LE4BHSL4DOUIK -j KUBE-SVC-X27LE4BHSL4DOUIK
+		-A KUBE-FW-GNZBNJ2PO5MGZ6GT -m comment --comment "ns2/svc2:p80 loadbalancer IP" -s 203.0.113.0/25 -j KUBE-EXT-GNZBNJ2PO5MGZ6GT
```
if you wanted this to be more-reviewable you could do the XLB->EXT renaming as a separate commit...

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1220,28 +1220,35 @@ func (proxier *Proxier) syncProxyRules() {
-				// This currently works for loadbalancers that preserves source ips.
-				// For loadbalancers which direct traffic to service NodePort, the firewall rules will not apply.
+		if len(svcInfo.LoadBalancerIPStrings()) > 0 && hasEndpoints {
+			// We could elide the firewall chain when loadBalancerSourceRanges
+			// is empty, but that adds code and makes the flow harder to reason
+			// about, for not much gain.
+			fwChain := svcInfo.firewallChainName
```
It doesn't add _much_ code, and I was planning to bring that commit back...

thockin replied:
It's OK as a followup, I guess.

thockin replied:
I added a WIP commit, PTAL

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1220,28 +1220,35 @@ func (proxier *Proxier) syncProxyRules() {
+			activeNATChains[fwChain] = true
+
+			for _, lbip := range svcInfo.LoadBalancerIPStrings() {
+				// The service firewall rules are created based on the
+				// loadBalancerSourceRanges field.  This only works for
+				// VIP-like loadbalancers that preserve source IPs.  For
+				// loadbalancers which direct traffic to service NodePort, the
+				// firewall rules will not apply.
```
In one of my not-currently-part-of-the-PR commits I had moved this comment to the top of the `if len(svcInfo.LoadBalancerIPStrings()) > 0` section (and reworded it a little), because the comment applies to this section as a whole, not just this loop.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1264,27 +1271,31 @@ func (proxier *Proxier) syncProxyRules() {
-					// generally, ip route rule was added to intercept request to loadbalancer vip from the
-					// loadbalancer's backend hosts. In this case, request will not hit the loadbalancer but loop back directly.
-					// Need to add the following rule to allow request on host.
+					// Generally, an IP route rule was added to intercept
+					// requests to the LB VIP from the loadbalancer's backend
+					// hosts. In this case, request will not hit the
+					// loadbalancer but loop back directly.  We need the
+					// following rule to allow requests from this host.
```
if you're going to re-wrap and re-capitalize this you should rewrite it to actually make sense too. :slightly_smiling_face:

danwinship commented on pkg/api/v1/service/util.go:
```diff
@@ -67,18 +67,18 @@ func GetLoadBalancerSourceRanges(service *v1.Service) (utilnet.IPNetSet, error)
 
-// RequestsOnlyLocalTraffic checks if service requests OnlyLocal traffic.
-func RequestsOnlyLocalTraffic(service *v1.Service) bool {
+// ExternalPolicyLocal checks if service requests OnlyLocal traffic.
```
"if service has Local external traffic policy"? and likewise below

aojea commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1066,56 +1080,73 @@ func (proxier *Proxier) syncProxyRules() {
+					"-m", "comment", "--comment", fmt.Sprintf(`"masquerade traffic for %s external destinations"`, svcNameString),
+					"-j", string(KubeMarkMasqChain))
+			} else {
+				// If we are proxying only locally, we can retain the source IP
+				// in most cases.
```
network-pod to host-network pod (in other node) requires masquerade

thockin replied:
I can't tell if you're saying I missed a case or just find the case weird?

aojea replied:
I was trying to remember, but this seems right, host-network pod in same node should not need masquerade

aojea commented:
the unit test failures are legit  https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109060/pull-kubernetes-unit/1508241640496566272

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -115,11 +115,11 @@ const sysctlBridgeCallIPTables = "net/bridge/bridge-nf-call-iptables"
-	serviceFirewallChainName utiliptables.Chain
-	serviceLBChainName       utiliptables.Chain
+	nameString             string
+	policyClusterChainName utiliptables.Chain
+	policyLocalChainName   utiliptables.Chain
```
I liked them at the time, but I didn't want to make you go change again, I'd rather get your PR in and move forward.  I'm note super attached to these names either, but the extra level of indirection I thought was clearer

thockin commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1308,16 +1310,16 @@ COMMIT
 // assertIPTablesRulesEqual asserts that the generated rules in result match the rules in
 // expected, ignoring irrelevant ordering differences.
 func assertIPTablesRulesEqual(t *testing.T, expected, result string) {
-	expected, err := sortIPTablesRules(expected)
```
I:

a) When the test fails it outputs the sorted result so you can cut-paste it into the test code
b) Fixed all the existing tests to be pre-sorted

This means that we'll not end up in cases where the string in the test doesn't match the diff ordering (which I found super weird).

thockin commented:
Good comments.  I wanted to put this on the radar, but I will shift attention to code reviews now and if I can get this in, great, if not, it can wait.

thockin committed:
Rename tests to avoid underscores

thockin committed:
kube-proxy: inline `args` in most cases

thockin committed:
Elide the -FW- chain when possible

This makes it epsilon harder to reason about but saves one chain
declaration and one rule per service-port usually.

thockin reviewed (commented):
Pushed with 2 WIP commits, will look at tests later.

aojea commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1701,21 +1701,23 @@ func TestClusterIPEndpointsJump(t *testing.T) {
 		-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
 		-A KUBE-SEP-SXIVWICOYRO3J4NJ -m comment --comment ns1/svc1:p80 -s 10.180.0.1 -j KUBE-MARK-MASQ
 		-A KUBE-SEP-SXIVWICOYRO3J4NJ -m comment --comment ns1/svc1:p80 -m tcp -p tcp -j DNAT --to-destination 10.180.0.1:80
-		-A KUBE-SVC-XPGD46QRK7WJZT7O -m comment --comment "ns1/svc1:p80 cluster IP" -m tcp -p tcp -d 172.30.0.41 --dport 80 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
```
xref https://github.com/kubernetes/kubernetes/pull/96959

aojea commented:
failures are legit in pull-kubernetes-unit — Job failed.                    
 https://github.com/kubernetes/kubernetes/pull/109060/commits/9b6e21920237514ea55570cf50184ebb83e2f235 description may be correct , the logic was modified recently to reduce the number of rules https://github.com/kubernetes/kubernetes/pull/96959

thockin committed:
kube-proxy: print line number for test failures

thockin commented:
Rebased on top of @andrewsykim #108691

thockin commented:
I'm going to drop the WIP commit (https://github.com/thockin/kubernetes/commit/498018936d729365304125e87a680254f1dbc05c) and we can discuss that separately.

Pushed

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -1307,18 +1308,33 @@ func Test_sortIPTablesRules(t *testing.T) {
+// getLine returns the line number of the caller, if possible.  This is useful in
+// tests with a large number of cases - when something goes wrong you can find
+// which case more easily.
+func getLine() int {
```
@aojea recently pointed me to [t.Helper()](https://pkg.go.dev/testing#T.Helper), which if you call it in a function, it causes future `t.Errorf()` etc calls in that function to print line numbers as though they had been called from the parent function instead...

thockin replied:
Righ, not so helpful with a table-test.  I used this line-num trick in REST tests, it helped a TON.

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -4393,6 +4414,7 @@ func TestInternalTrafficPolicyE2E(t *testing.T) {
 	}{
 		{
 			name:                  "internalTrafficPolicy is cluster",
+			line:                  getLine(),
```
ah, `t.Helper()` wouldn't help quite as much with this, although because all the tests use `t.Run(tc.name, ...` now you can just search for the name string from the failure message rather than the line number

thockin replied:
The name string is also not-great because Go helpfully (!!) converts " " to "_" so I can't cut-paste it.  Also, we have cases in different tests which use the same name.

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1066,56 +1076,73 @@ func (proxier *Proxier) syncProxyRules() {
-			// is implemented
-			if proxier.localDetector.IsImplemented() {
+			if !svcInfo.NodeLocalExternal() || !proxier.localDetector.IsImplemented() {
+				// If we are proxying globally (or if we can't tell local from
```
Maybe fix up this comment? "Proxying globally" is outdated terminology. (Likewise "proxying locally" below.)

thockin replied:
ACK

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1315,7 +1308,9 @@ func (proxier *Proxier) syncProxyRules() {
 						"-s", loopback, "-j", string(KubeMarkMasqChain))
 				}
-				// Jump to the service chain.
+				// Jump to the external desination chain.  For better or for
```
typo "desination"

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1066,56 +1076,73 @@ func (proxier *Proxier) syncProxyRules() {
+				// If we are proxying only locally, we can retain the source IP
+				// in most cases.
+
+				// Treat all locally-originated pod -> external destination
+				// traffic as a special-case.  It is subject to neither form of
+				// traffic policy, which simulates going up-and-out to an
+				// external load-balancer and coming back in.
+				proxier.natRules.Write(
+					"-A", string(externalTrafficChain),
+					"-m", "comment", "--comment", fmt.Sprintf(`"masquerade traffic for %s external destinations"`, svcNameString),
 					proxier.localDetector.IfLocal(),
 					"-j", string(policyClusterChain))
```
That `--comment` is wrong.

(True confession: I noticed this while examining the iptables diffs in proxier_test.go!)

thockin replied:
ACK

danwinship commented on pkg/proxy/iptables/proxier_test.go:
```diff
@@ -458,7 +458,7 @@ func parseIPTablesData(ruleData string) (map[string][]string, error) {
 	return tables, nil
 }
 
-func Test_parseIPTablesData(t *testing.T) {
+func TestParseIPTablesData(t *testing.T) {
```
This is a not-uncommon convention for testing functions with initial-lowercase names:

    danw@p1:kubernetes> cat $(find pkg -name \*_test.go) | grep -c '^func Test_[a-z]'
    65

thockin replied:
I see.  It's still ... ugh  I wish the norm was ALWAYS Test_.

Are we going to argue this?  I can partially revert it if you feel strongly :)

danwinship reviewed (commented):
I was going to /lgtm /hold, but you really ought to fix the `--comment` on the short-circuit rule.
I'm OK with it merging without any of the other suggested changes (though I'm even more OK with it merging _with_ the suggested changes...)

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -115,11 +115,11 @@ const sysctlBridgeCallIPTables = "net/bridge/bridge-nf-call-iptables"
-	serviceLBChainName       utiliptables.Chain
+	nameString             string
+	policyClusterChainName utiliptables.Chain
+	policyLocalChainName   utiliptables.Chain
```
good change, will do

thockin commented:
Pushed with all fixes except Test_ :)

thockin commented:
Hold until #108691 merges

thockin commented:
Somebody LGTM me before the clock runs out?

danwinship commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1042,80 +1045,108 @@ func (proxier *Proxier) syncProxyRules() {
-				"-m", "addrtype", "--src-type", "LOCAL",
-				"-j", string(policyClusterChain))
+			if !svcInfo.ExternalPolicyLocal() || !proxier.localDetector.IsImplemented() {
+				// If we have are using non-local endpoints (or if we can't
```
oh, since you have to re-push anyway:

```suggestion
				// If we are using non-local endpoints (or if we can't
```

thockin replied:
ACK

thockin commented:
Rebased and pushed

andrewsykim commented:
/lgtm

thockin commented:
/retest

aojea commented:
/retest

danwinship commented:
/retest-required

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1042,80 +1045,107 @@ func (proxier *Proxier) syncProxyRules() {
-				"-m", "comment", "--comment", fmt.Sprintf(`"route LOCAL traffic for %s LB IP to service chain"`, svcNameString),
-				"-m", "addrtype", "--src-type", "LOCAL",
-				"-j", string(policyClusterChain))
+			if !svcInfo.ExternalPolicyLocal() || !proxier.localDetector.IsImplemented() {
+				// If we are using non-local endpoints (or if we can't tell
+				// local from non-local because there's no local-detector), we
+				// need to masquerade, in case we cross nodes.
+				proxier.natRules.Write(
+					"-A", string(externalTrafficChain),
+					"-m", "comment", "--comment", fmt.Sprintf(`"masquerade traffic for %s external destinations"`, svcNameString),
+					"-j", string(KubeMarkMasqChain))
```
Good catch!  Pushing update

thockin replied:
Try now?

danwinship commented:
/lgtm
passes rebased 107471 tests locally
